warehouse/slave.go

```diff
@@ -189,126 +207,170 @@ func (job *Payload) getDiscardsTable() string {
-	// take the first staging file id in upload
-	// TODO: support multiple staging files in one upload
-	stagingFileId := jobRun.job.StagingFileID
-
-	loadFileOutputChan := make(chan loadFileUploadOutput, len(jobRun.outputFileWritersMap))
-	loadFileUploadTimer := jobRun.timerStat("load_file_upload_time")
-	uploadJobChan := make(chan *loadFileUploadJob, len(jobRun.outputFileWritersMap))
-	// close chan to avoid memory leak ranging over it
-	defer close(uploadJobChan)
-	uploadErrorChan := make(chan error, numLoadFileUploadWorkers)
-	ctx, cancel := context.WithCancel(ctx)
-	defer cancel()
-	for i := 0; i < numLoadFileUploadWorkers; i++ {
-		go func(ctx context.Context) {
-			for uploadJob := range uploadJobChan {
+		return nil, fmt.Errorf("creating uploader: %w", err)
+	}
+
+	type result struct {
+		uploadOutput loadFileUploadOutput
+		err          error
+	}
+
+	type uploadJob struct {
+		tableName  string
+		outputFile encoding.LoadFileWriter
+	}
+
+	var output []loadFileUploadOutput
+	var totalUploadTime time.Duration
+	var mu sync.Mutex
+
+	defer func() {
+		if totalUploadTime > 0 {
+			jr.timerStat("load_file_upload_time").SendTiming(totalUploadTime)
+		}
+	}()
+
+	generate := func() <-chan *uploadJob {
+		jobStream := make(chan *uploadJob, len(jr.outputFileWritersMap))
+		go func() {
+			defer close(jobStream)
+
+			for tableName, loadFile := range jr.outputFileWritersMap {
 				select {
 				case <-ctx.Done():
-					pkgLogger.Debugf("context is cancelled, stopped processing load file for staging file ids %s ", stagingFileId)
-					return // stop further processing
+					return
 				default:
-					tableName := uploadJob.tableName
-					loadFileUploadStart := time.Now()
-					uploadOutput, err := jobRun.uploadLoadFileToObjectStorage(ctx, uploader, uploadJob.outputFile, tableName)
-					if err != nil {
-						uploadErrorChan <- err
-						return
-					}
-					loadFileUploadTimer.Since(loadFileUploadStart)
-					loadFileStats, err := os.Stat(uploadJob.outputFile.GetLoadFile().Name())
-					if err != nil {
-						uploadErrorChan <- err
-						return
-					}
-					loadFileOutputChan <- loadFileUploadOutput{
-						TableName:             tableName,
-						Location:              uploadOutput.Location,
-						ContentLength:         loadFileStats.Size(),
-						TotalRows:             jobRun.tableEventCountMap[tableName],
-						StagingFileID:         stagingFileId,
-						DestinationRevisionID: job.DestinationRevisionID,
-						UseRudderStorage:      job.UseRudderStorage,
+					jobStream <- &uploadJob{
+						tableName:  tableName,
+						outputFile: loadFile,
 					}
 				}
 			}
-		}(ctx)
+		}()
+		return jobStream
 	}
-	// Create upload jobs
-	go func() {
-		for tableName, loadFile := range jobRun.outputFileWritersMap {
-			uploadJobChan <- &loadFileUploadJob{tableName: tableName, outputFile: loadFile}
+
+	uploadLoadFile := func(uploader filemanager.FileManager, uploadFile encoding.LoadFileWriter, tableName string) (filemanager.UploadedFile, error) {
```

Comment:
Since this is a lambda we can simplify this a bit and not pass `uploader`. Same thing you did on `process` down below.